File: clients/rn/src/server/zoneBankSync.ts
/**
 * zoneBankSync — cross-device sync of the zone bank.
 *
 * The bank (the address book of zones) is sovereign account state mirrored on
 * the user's home zone. Each authenticated device pulls it on connect and
 * pushes its merged view back, so every device the user signs in on converges
 * to the union of their zones.
 *
 * What syncs: the `zones[]` address book only. What does NOT:
 *   • held relays  — transport credentials (natsUser/natsPass) are per-device
 *     secrets, pinned via TOFU from invites. Never travel through the bank blob.
 *   • zone passwords — per-device secure storage (@wyrd_zone_pw_*), §4.4.
 *
 * Merge is per-entry last-write-wins, keyed by zoneId. "Newest" = the larger of
 * (lastUsedAt, addedAt) — a zone touched more recently on any device wins. The
 * merge is symmetric and idempotent: pull → merge newest-wins into local → push
 * the merged superset. Two devices that both run this converge.
 *
 * The server (McpNatsHandler account.zonebank.{get,put}) is a dumb blob store
 * keyed by the caller's account; the LWW logic lives here on the client.
 */
import { useZoneBankStore, type ZoneBankEntry } from '../state/zoneBankStore';
import type { NatsResult } from './NatsServerClient';

/** The slice of NatsServerClient this module needs — keeps it test-mockable. */
export interface ZoneBankSyncClient {
  getZoneBank(): Promise<NatsResult<{ bank: string | null; updatedAt: number }>>;
  putZoneBank(bankJson: string, updatedAt: number): Promise<NatsResult<{ updatedAt: number }>>;
}

/** Recency stamp used for per-entry LWW. */
function entryTs(z: ZoneBankEntry): number {
  return Math.max(z.lastUsedAt ?? 0, z.addedAt ?? 0);
}

/**
 * Merge a remote zones[] array into the local store, per-entry newest-wins.
 * Relay credentials are NOT carried in the blob, so a synced entry may name
 * relayUrls this device hasn't pinned yet — that's fine; relaysForZone() falls
 * back to all held relays. Returns the number of entries added or updated.
 */
export function mergeRemoteZones(remote: ZoneBankEntry[]): number {
  const store = useZoneBankStore.getState();
  let changed = 0;
  for (const r of remote) {
    if (!r || typeof r.zoneId !== 'string' || !r.zoneId) continue;
    const local = store.getZone(r.zoneId);
    if (!local || entryTs(r) > entryTs(local)) {
      // addOrUpdateZone unions relayUrls + keeps local addedAt; for a brand-new
      // remote zone it inserts wholesale. Both are correct for a newest-wins
      // remote entry.
      store.addOrUpdateZone({
        zoneId: r.zoneId,
        displayName: r.displayName ?? r.zoneId,
        relayUrls: Array.isArray(r.relayUrls) ? r.relayUrls : [],
        username: r.username ?? '',
        homeZone: r.homeZone,
        lastUsedAt: r.lastUsedAt,
      });
      changed++;
    }
  }
  return changed;
}

/** Serialize the current local bank for upload — zones only, no secrets. */
export function serializeBank(): string {
  return JSON.stringify(useZoneBankStore.getState().zones);
}

function parseRemote(bank: string | null): ZoneBankEntry[] {
  if (!bank) return [];
  try {
    const parsed = JSON.parse(bank);
    return Array.isArray(parsed) ? (parsed as ZoneBankEntry[]) : [];
  } catch {
    return [];
  }
}

export interface SyncResult {
  ok: boolean;
  pulled: number; // entries merged in from the server
  pushed: boolean; // whether we uploaded the merged bank
  error?: string;
}

/**
 * Full sync against the connected zone: pull the server's bank, merge it
 * newest-wins into the local store, then push the merged superset back so the
 * server (and thus the user's other devices) converges. Best-effort — a sync
 * failure never blocks the session; the local bank remains authoritative on
 * this device.
 *
 * @param now monotonic-ish timestamp for the push stamp (Date.now()); passed in
 *            so callers/tests control it and the module stays pure.
 */
export async function syncZoneBank(
  client: ZoneBankSyncClient,
  now: number,
): Promise<SyncResult> {
  const got = await client.getZoneBank();
  if (!got.ok) {
    return { ok: false, pulled: 0, pushed: false, error: got.error };
  }
  const remote = parseRemote(got.data?.bank ?? null);
  const pulled = mergeRemoteZones(remote);

  // Push the merged view. We push even when pulled===0: this device may hold
  // zones the server hasn't seen yet (first sync from a device that onboarded
  // offline). Skip only when the local bank is empty AND nothing came down.
  const localZones = useZoneBankStore.getState().zones;
  if (localZones.length === 0) {
    return { ok: true, pulled, pushed: false };
  }
  const put = await client.putZoneBank(serializeBank(), now);
  if (!put.ok) {
    return { ok: true, pulled, pushed: false, error: put.error };
  }
  return { ok: true, pulled, pushed: true };
}

File: clients/kmp/shared/src/commonMain/kotlin/org/wyrdsekai/app/network/ZoneBankSync.kt
package org.wyrdsekai.app.network

/**
 * ZoneBankSync — cross-device sync of the zone bank
 * KMP parity (P5) with the RN `zoneBankSync` module.
 *
 * The bank (the address book of zones) is sovereign account state mirrored on
 * the user's home zone. Each authenticated device pulls it on connect and pushes
 * its merged view back, so every device converges on the union of the user's
 * zones. Only the zones[] address book travels — held-relay credentials and zone
 * passwords are per-device secrets that never enter the blob (§4.4).
 *
 * Merge is per-entry last-write-wins by zoneId, "newest" = max(lastUsedAt,
 * addedAt). Symmetric and idempotent: pull → merge newest-wins → push the merged
 * superset. The server is a dumb blob store keyed by account; the LWW logic lives
 * here, identical to RN so the two clients interoperate.
 */

/** Server's stored bank + its stamp (null bank = nothing stored yet). */
data class ZoneBankFetch(val bank: String?, val updatedAt: Long)

/** The slice of the NATS client this module needs — keeps it test-mockable. */
interface ZoneBankSyncClient {
    /** @return the stored bank, or null on transport/auth failure. */
    suspend fun getZoneBank(): ZoneBankFetch?
    /** @return true on success. */
    suspend fun putZoneBank(bankJson: String, updatedAt: Long): Boolean
}

data class SyncResult(
    val ok: Boolean,
    val pulled: Int,    // entries merged in from the server
    val pushed: Boolean, // whether we uploaded the merged bank
    val error: String? = null,
)

object ZoneBankSync {

    private fun entryTs(z: ZoneBankEntry): Long = maxOf(z.lastUsedAt ?: 0L, z.addedAt)

    /**
     * Merge a remote zones list into [bank], per-entry newest-wins. Returns the
     * number of entries added or updated. Relay credentials are not in the blob,
     * so a synced entry may name relayUrls this device hasn't pinned yet — fine;
     * relaysForZone falls back to all held relays.
     */
    fun mergeRemoteZones(bank: ZoneBank, remote: List<ZoneBankEntry>): Int {
        var changed = 0
        for (r in remote) {
            if (r.zoneId.isBlank()) continue
            val local = bank.getZone(r.zoneId)
            if (local == null || entryTs(r) > entryTs(local)) {
                bank.addOrUpdateZone(r)
                changed++
            }
        }
        return changed
    }

    private fun parseRemote(bank: ZoneBank, raw: String?): List<ZoneBankEntry> {
        if (raw.isNullOrBlank()) return emptyList()
        return try {
            // Reuse ZoneBank's decoder by round-tripping through a scratch bank.
            ZoneBank().also { it.load(null, raw) }.zones
        } catch (_: Exception) {
            emptyList()
        }
    }

    /**
     * Full sync against the connected zone: pull → merge newest-wins into [bank]
     * → push the merged superset so the server (and the user's other devices)
     * converges. Best-effort — a failure never blocks the session; the local bank
     * stays authoritative on this device.
     */
    suspend fun syncZoneBank(bank: ZoneBank, client: ZoneBankSyncClient, now: Long): SyncResult {
        val fetched = client.getZoneBank()
            ?: return SyncResult(ok = false, pulled = 0, pushed = false, error = "zonebank get failed")
        val pulled = mergeRemoteZones(bank, parseRemote(bank, fetched.bank))

        // Push even when pulled==0: this device may hold zones the server hasn't
        // seen. Skip only when the local bank is empty AND nothing came down.
        if (bank.zones.isEmpty()) {
            return SyncResult(ok = true, pulled = pulled, pushed = false)
        }
        val ok = client.putZoneBank(bank.serializeZones(), now)
        return SyncResult(ok = true, pulled = pulled, pushed = ok, error = if (ok) null else "zonebank put failed")
    }
}
